File: backend/app/services/cost_estimator.py
"""
Cost-to-Comply Estimator & ROI Calculator

Hard-coded fixture pricing database and energy savings calculations.
No external API calls — all data is embedded.
"""

from dataclasses import dataclass, field
from typing import List, Optional

# ── Fixture Pricing Database ─────────────────────────────────────────────────
# Realistic 2024-2025 pricing for common commercial LED fixtures.
# Each entry: (fixture_type, min_price, max_price, typical_replacement_wattage)

@dataclass
class FixturePricing:
    fixture_type: str
    min_price: float
    max_price: float
    avg_price: float
    typical_wattage: float  # typical compliant replacement wattage
    dlc_listed: bool        # whether DLC QPL listed (rebate eligible)


FIXTURE_PRICING_DB: dict[str, FixturePricing] = {}

_RAW_PRICING = [
    # (type, min, max, typical_replacement_W, dlc_listed)
    ("LED Troffer",         45,   120,   30,  True),
    ("LED Panel",           40,   110,   32,  True),
    ("LED High Bay",        80,   250,  150,  True),
    ("LED Low Bay",         55,   150,   75,  True),
    ("LED Downlight",       15,    65,   12,  True),
    ("LED Strip",           10,    45,   14,  False),
    ("LED Wall Pack",       50,   180,   30,  True),
    ("LED Flood Light",     60,   220,   80,  True),
    ("LED Area Light",      90,   350,  100,  True),
    ("LED Bollard",         80,   250,   15,  True),
    ("LED Canopy",          65,   200,   45,  True),
    ("LED Parking Garage",  70,   220,   40,  True),
    ("LED Stairwell",       35,   100,   20,  True),
    ("LED Vapor Tight",     45,   130,   35,  True),
    ("LED Track Light",     20,    80,   10,  False),
    ("LED Pendant",         55,   250,   25,  True),
    ("LED Sconce",          25,    90,   10,  False),
    ("LED Emergency",       30,    85,    5,  False),
    ("LED Exit Sign",       20,    60,    2,  False),
    ("Other",               40,   150,   30,  False),
]

for _type, _min, _max, _watt, _dlc in _RAW_PRICING:
    FIXTURE_PRICING_DB[_type] = FixturePricing(
        fixture_type=_type,
        min_price=_min,
        max_price=_max,
        avg_price=round((_min + _max) / 2, 2),
        typical_wattage=_watt,
        dlc_listed=_dlc,
    )

# ── Energy defaults ──────────────────────────────────────────────────────────

DEFAULT_UTILITY_RATE = 0.12       # $/kWh
DEFAULT_ANNUAL_HOURS = 4380       # hours/year (commercial standard)
TYPICAL_REBATE_MIN = 20.0         # $/fixture
TYPICAL_REBATE_MAX = 50.0         # $/fixture


# ── Data classes ─────────────────────────────────────────────────────────────

@dataclass
class FixtureCostEstimate:
    fixture_id: str
    fixture_type: str
    fixture_name: str
    current_wattage: float
    replacement_wattage: float
    quantity: int
    replacement_cost_min: float
    replacement_cost_max: float
    replacement_cost_avg: float
    annual_energy_current_kwh: float
    annual_energy_replacement_kwh: float
    annual_energy_savings_kwh: float
    annual_cost_current: float
    annual_cost_replacement: float
    annual_savings: float
    payback_years: Optional[float]
    dlc_listed: bool
    rebate_eligible: bool
    estimated_rebate_min: float
    estimated_rebate_max: float


@dataclass
class SiteROISummary:
    total_fixtures: int
    failing_fixtures: int
    total_replacement_cost_min: float
    total_replacement_cost_max: float
    total_replacement_cost_avg: float
    total_annual_energy_current_kwh: float
    total_annual_energy_replacement_kwh: float
    total_annual_energy_savings_kwh: float
    total_annual_cost_current: float
    total_annual_cost_replacement: float
    total_annual_savings: float
    avg_payback_years: Optional[float]
    rebate_eligible_count: int
    estimated_total_rebate_min: float
    estimated_total_rebate_max: float
    fixture_estimates: List[FixtureCostEstimate]


# ── Calculator functions ─────────────────────────────────────────────────────

def estimate_fixture_cost(
    fixture_id: str,
    fixture_type: str,
    fixture_name: str,
    current_wattage: float,
    quantity: int,
    utility_rate: float = DEFAULT_UTILITY_RATE,
    annual_hours: float = DEFAULT_ANNUAL_HOURS,
) -> FixtureCostEstimate:
    """Estimate replacement cost and ROI for a single failing fixture."""
    pricing = FIXTURE_PRICING_DB.get(fixture_type, FIXTURE_PRICING_DB["Other"])

    replacement_wattage = pricing.typical_wattage
    qty = max(quantity, 1)

    # Costs
    cost_min = pricing.min_price * qty
    cost_max = pricing.max_price * qty
    cost_avg = pricing.avg_price * qty

    # Energy (kWh/yr)
    current_kwh = (current_wattage * qty * annual_hours) / 1000.0
    replacement_kwh = (replacement_wattage * qty * annual_hours) / 1000.0
    savings_kwh = current_kwh - replacement_kwh

    # Dollar amounts
    annual_cost_current = current_kwh * utility_rate
    annual_cost_replacement = replacement_kwh * utility_rate
    annual_savings = savings_kwh * utility_rate

    # Payback
    payback = None
    if annual_savings > 0:
        payback = round(cost_avg / annual_savings, 2)

    # Rebate
    dlc = pricing.dlc_listed
    rebate_min = TYPICAL_REBATE_MIN * qty if dlc else 0.0
    rebate_max = TYPICAL_REBATE_MAX * qty if dlc else 0.0

    return FixtureCostEstimate(
        fixture_id=fixture_id,
        fixture_type=fixture_type,
        fixture_name=fixture_name or fixture_type,
        current_wattage=current_wattage,
        replacement_wattage=replacement_wattage,
        quantity=qty,
        replacement_cost_min=round(cost_min, 2),
        replacement_cost_max=round(cost_max, 2),
        replacement_cost_avg=round(cost_avg, 2),
        annual_energy_current_kwh=round(current_kwh, 2),
        annual_energy_replacement_kwh=round(replacement_kwh, 2),
        annual_energy_savings_kwh=round(savings_kwh, 2),
        annual_cost_current=round(annual_cost_current, 2),
        annual_cost_replacement=round(annual_cost_replacement, 2),
        annual_savings=round(annual_savings, 2),
        payback_years=payback,
        dlc_listed=dlc,
        rebate_eligible=dlc,
        estimated_rebate_min=round(rebate_min, 2),
        estimated_rebate_max=round(rebate_max, 2),
    )


def calculate_site_roi(
    failing_fixtures: List[dict],
    total_fixture_count: int,
    utility_rate: float = DEFAULT_UTILITY_RATE,
    annual_hours: float = DEFAULT_ANNUAL_HOURS,
) -> SiteROISummary:
    """
    Calculate ROI summary for all failing fixtures at a site.

    failing_fixtures: list of dicts with keys:
        fixture_id, fixture_type, fixture_name, wattage, quantity
    """
    estimates: List[FixtureCostEstimate] = []
    for fx in failing_fixtures:
        est = estimate_fixture_cost(
            fixture_id=fx["fixture_id"],
            fixture_type=fx["fixture_type"],
            fixture_name=fx.get("fixture_name", ""),
            current_wattage=fx.get("wattage", 0) or 0,
            quantity=fx.get("quantity", 1),
            utility_rate=utility_rate,
            annual_hours=annual_hours,
        )
        estimates.append(est)

    total_cost_min = sum(e.replacement_cost_min for e in estimates)
    total_cost_max = sum(e.replacement_cost_max for e in estimates)
    total_cost_avg = sum(e.replacement_cost_avg for e in estimates)
    total_current_kwh = sum(e.annual_energy_current_kwh for e in estimates)
    total_repl_kwh = sum(e.annual_energy_replacement_kwh for e in estimates)
    total_savings_kwh = sum(e.annual_energy_savings_kwh for e in estimates)
    total_cost_current = sum(e.annual_cost_current for e in estimates)
    total_cost_repl = sum(e.annual_cost_replacement for e in estimates)
    total_savings = sum(e.annual_savings for e in estimates)
    rebate_count = sum(1 for e in estimates if e.rebate_eligible)
    rebate_min = sum(e.estimated_rebate_min for e in estimates)
    rebate_max = sum(e.estimated_rebate_max for e in estimates)

    avg_payback = None
    if total_savings > 0:
        avg_payback = round(total_cost_avg / total_savings, 2)

    return SiteROISummary(
        total_fixtures=total_fixture_count,
        failing_fixtures=len(estimates),
        total_replacement_cost_min=round(total_cost_min, 2),
        total_replacement_cost_max=round(total_cost_max, 2),
        total_replacement_cost_avg=round(total_cost_avg, 2),
        total_annual_energy_current_kwh=round(total_current_kwh, 2),
        total_annual_energy_replacement_kwh=round(total_repl_kwh, 2),
        total_annual_energy_savings_kwh=round(total_savings_kwh, 2),
        total_annual_cost_current=round(total_cost_current, 2),
        total_annual_cost_replacement=round(total_cost_repl, 2),
        total_annual_savings=round(total_savings, 2),
        avg_payback_years=avg_payback,
        rebate_eligible_count=rebate_count,
        estimated_total_rebate_min=round(rebate_min, 2),
        estimated_total_rebate_max=round(rebate_max, 2),
        fixture_estimates=estimates,
    )


def get_pricing_for_type(fixture_type: str) -> FixturePricing:
    """Look up pricing for a fixture type."""
    return FIXTURE_PRICING_DB.get(fixture_type, FIXTURE_PRICING_DB["Other"])


def get_all_pricing() -> List[FixturePricing]:
    """Return all pricing entries."""
    return list(FIXTURE_PRICING_DB.values())

File: frontend/lib/compliance/costEstimator.ts
/**
 * Client-side Cost-to-Comply Estimator & ROI Calculator
 * Mirrors backend/app/services/cost_estimator.py
 */

import type { Fixture, ComplianceStatus } from "./types"

// ── Fixture Pricing Database ────────────────────────────────────────────────

export interface FixturePricing {
  fixtureType: string
  minPrice: number
  maxPrice: number
  avgPrice: number
  typicalWattage: number
  dlcListed: boolean
}

const RAW_PRICING: [string, number, number, number, boolean][] = [
  ["LED Troffer",         45,   120,   30,  true],
  ["LED Panel",           40,   110,   32,  true],
  ["LED High Bay",        80,   250,  150,  true],
  ["LED Low Bay",         55,   150,   75,  true],
  ["LED Downlight",       15,    65,   12,  true],
  ["LED Strip",           10,    45,   14,  false],
  ["LED Wall Pack",       50,   180,   30,  true],
  ["LED Flood Light",     60,   220,   80,  true],
  ["LED Area Light",      90,   350,  100,  true],
  ["LED Bollard",         80,   250,   15,  true],
  ["LED Canopy",          65,   200,   45,  true],
  ["LED Parking Garage",  70,   220,   40,  true],
  ["LED Stairwell",       35,   100,   20,  true],
  ["LED Vapor Tight",     45,   130,   35,  true],
  ["LED Track Light",     20,    80,   10,  false],
  ["LED Pendant",         55,   250,   25,  true],
  ["LED Sconce",          25,    90,   10,  false],
  ["LED Emergency",       30,    85,    5,  false],
  ["LED Exit Sign",       20,    60,    2,  false],
  ["Other",               40,   150,   30,  false],
]

export const FIXTURE_PRICING_DB: Record<string, FixturePricing> = {}
for (const [type, min, max, watt, dlc] of RAW_PRICING) {
  FIXTURE_PRICING_DB[type] = {
    fixtureType: type,
    minPrice: min,
    maxPrice: max,
    avgPrice: Math.round(((min + max) / 2) * 100) / 100,
    typicalWattage: watt,
    dlcListed: dlc,
  }
}

// ── Defaults ────────────────────────────────────────────────────────────────

export const DEFAULT_UTILITY_RATE = 0.12    // $/kWh
export const DEFAULT_ANNUAL_HOURS = 4380    // hrs/yr (commercial)
export const TYPICAL_REBATE_MIN = 20        // $/fixture
export const TYPICAL_REBATE_MAX = 50        // $/fixture

// ── Types ───────────────────────────────────────────────────────────────────

export interface FixtureCostEstimate {
  fixtureId: string
  fixtureType: string
  fixtureName: string
  currentWattage: number
  replacementWattage: number
  quantity: number
  replacementCostMin: number
  replacementCostMax: number
  replacementCostAvg: number
  annualEnergyCurrent: number   // kWh
  annualEnergyReplacement: number
  annualEnergySavings: number
  annualCostCurrent: number     // $
  annualCostReplacement: number
  annualSavings: number
  paybackYears: number | null
  dlcListed: boolean
  rebateEligible: boolean
  estimatedRebateMin: number
  estimatedRebateMax: number
}

export interface SiteROISummary {
  totalFixtures: number
  failingFixtures: number
  totalReplacementCostMin: number
  totalReplacementCostMax: number
  totalReplacementCostAvg: number
  totalAnnualEnergyCurrent: number
  totalAnnualEnergyReplacement: number
  totalAnnualEnergySavings: number
  totalAnnualCostCurrent: number
  totalAnnualCostReplacement: number
  totalAnnualSavings: number
  avgPaybackYears: number | null
  rebateEligibleCount: number
  estimatedTotalRebateMin: number
  estimatedTotalRebateMax: number
  fixtureEstimates: FixtureCostEstimate[]
}

// ── Calculator ──────────────────────────────────────────────────────────────

function r2(n: number): number {
  return Math.round(n * 100) / 100
}

export function estimateFixtureCost(
  fixtureId: string,
  fixtureType: string,
  fixtureName: string,
  currentWattage: number,
  quantity: number,
  utilityRate = DEFAULT_UTILITY_RATE,
  annualHours = DEFAULT_ANNUAL_HOURS,
): FixtureCostEstimate {
  const pricing = FIXTURE_PRICING_DB[fixtureType] ?? FIXTURE_PRICING_DB["Other"]
  const qty = Math.max(quantity, 1)
  const replW = pricing.typicalWattage

  const costMin = pricing.minPrice * qty
  const costMax = pricing.maxPrice * qty
  const costAvg = pricing.avgPrice * qty

  const currentKwh = (currentWattage * qty * annualHours) / 1000
  const replKwh = (replW * qty * annualHours) / 1000
  const savingsKwh = currentKwh - replKwh

  const annualCostCurr = currentKwh * utilityRate
  const annualCostRepl = replKwh * utilityRate
  const annualSavings = savingsKwh * utilityRate

  const payback = annualSavings > 0 ? r2(costAvg / annualSavings) : null

  const dlc = pricing.dlcListed
  const rebateMin = dlc ? TYPICAL_REBATE_MIN * qty : 0
  const rebateMax = dlc ? TYPICAL_REBATE_MAX * qty : 0

  return {
    fixtureId,
    fixtureType,
    fixtureName: fixtureName || fixtureType,
    currentWattage,
    replacementWattage: replW,
    quantity: qty,
    replacementCostMin: r2(costMin),
    replacementCostMax: r2(costMax),
    replacementCostAvg: r2(costAvg),
    annualEnergyCurrent: r2(currentKwh),
    annualEnergyReplacement: r2(replKwh),
    annualEnergySavings: r2(savingsKwh),
    annualCostCurrent: r2(annualCostCurr),
    annualCostReplacement: r2(annualCostRepl),
    annualSavings: r2(annualSavings),
    paybackYears: payback,
    dlcListed: dlc,
    rebateEligible: dlc,
    estimatedRebateMin: r2(rebateMin),
    estimatedRebateMax: r2(rebateMax),
  }
}

export function calculateSiteROI(
  failingFixtures: { fixtureId: string; fixtureType: string; fixtureName: string; wattage: number; quantity: number }[],
  totalFixtureCount: number,
  utilityRate = DEFAULT_UTILITY_RATE,
  annualHours = DEFAULT_ANNUAL_HOURS,
): SiteROISummary {
  const estimates = failingFixtures.map((fx) =>
    estimateFixtureCost(fx.fixtureId, fx.fixtureType, fx.fixtureName, fx.wattage, fx.quantity, utilityRate, annualHours)
  )

  const sum = (fn: (e: FixtureCostEstimate) => number) => r2(estimates.reduce((a, e) => a + fn(e), 0))

  const totalCostAvg = sum((e) => e.replacementCostAvg)
  const totalSavings = sum((e) => e.annualSavings)

  return {
    totalFixtures: totalFixtureCount,
    failingFixtures: estimates.length,
    totalReplacementCostMin: sum((e) => e.replacementCostMin),
    totalReplacementCostMax: sum((e) => e.replacementCostMax),
    totalReplacementCostAvg: totalCostAvg,
    totalAnnualEnergyCurrent: sum((e) => e.annualEnergyCurrent),
    totalAnnualEnergyReplacement: sum((e) => e.annualEnergyReplacement),
    totalAnnualEnergySavings: sum((e) => e.annualEnergySavings),
    totalAnnualCostCurrent: sum((e) => e.annualCostCurrent),
    totalAnnualCostReplacement: sum((e) => e.annualCostReplacement),
    totalAnnualSavings: totalSavings,
    avgPaybackYears: totalSavings > 0 ? r2(totalCostAvg / totalSavings) : null,
    rebateEligibleCount: estimates.filter((e) => e.rebateEligible).length,
    estimatedTotalRebateMin: sum((e) => e.estimatedRebateMin),
    estimatedTotalRebateMax: sum((e) => e.estimatedRebateMax),
    fixtureEstimates: estimates,
  }
}

export function getPricing(fixtureType: string): FixturePricing {
  return FIXTURE_PRICING_DB[fixtureType] ?? FIXTURE_PRICING_DB["Other"]
}
